### PR TITLE
WIP Support dataset streaming

### PR DESCRIPTION
**Pull request by dlwh**
DRAFT

hacky WIP to unblock #126 . I hate almost everything about this, but it seems to function. I haven't actually tested that it does what we want for multidevice, and I have no idea if it's reasonable for multinode. It also documents all the mines I stepped on as I was doing this, which is probably helpful for us as we figure out what we actually want a solution to look like.

### CODE REVIEW

File: src/corpora/auto.py [hunk 1]
```diff
@@ -221,3 +242,85 @@ def tokenize_and_group(example: Dict[str, str]) -> Dict[str, List[int]]:
...
 }
+
+
+# more hax: there's a method in datasets called iterable_dataset that mixes in torch IterableDataset like this, but it
+# creates an instance of a local class, and the local class isn't pickleable, so... we'll make our own mixed in version
+class TorchIterableDataset(IterableDataset, torch.utils.data.IterableDataset):
+    pass
+
+
+def _ensure_format(dataset: IterableDatasetDict) -> IterableDatasetDict:
```
Comment (siddk) [inline src/corpora/auto.py:253]: Oof

File: src/corpora/auto.py [hunk 2]
```diff
@@ -58,103 +84,98 @@ def get_auto_dataset(
...
-    overwatch.info(f"Auto-Batching Dataset via Multiprocessing with `{preprocessing_num_proc}` threads...")
-
-    # Create Post-Chunking Cache Paths
-    post_chunking_cache_files = {
-        k: str(paths["preprocessed"] / dataset_id / "preprocessing" / "chunking" / f"{k}-stride={stride}-chunked.hf")
-        for k in dataset
-    }
-    # Create Parent Path of Cache Files
-    (paths["preprocessed"] / dataset_id / "preprocessing" / "chunking").mkdir(parents=True, exist_ok=True)
+    if streaming:
```
Comment (siddk) [inline src/corpora/auto.py:125]: Ok at this point I'm wondering if we should drop HF Datasets?

File: src/corpora/auto.py [hunk 3]
```diff
@@ -29,20 +34,41 @@ def get_auto_dataset(
...
     dataset = datasets.load_dataset(
-        dataset_id, name=dataset_name, cache_dir=str(paths["dataset"]), keep_in_memory=True
+        dataset_id, name=dataset_name, cache_dir=str(paths["dataset"]), keep_in_memory=not streaming,
+        streaming=streaming,
     )
 
+    # while HF datasets has a map method for both normal Dataset and streaming IterableDataset, the IterableDataset
+    # version doesn't actually work with Trainer so we have to wrap it in this StreamedDataset nonsense.
+    # The reasons are:
+    # 1. IterableDataset transforms are pickled using pickle (not dill or something reasonable), so you have to be super
```
Comment (siddk) [inline src/corpora/auto.py:53]: Can we upstream this as an issue in HF Datasets?

File: src/corpora/auto.py [hunk 4]
```diff
@@ -29,20 +34,41 @@ def get_auto_dataset(
...
+    # careful about what you pass to map. This is mostly Python's fault.
+    # 2. IterableDataset doesn't have a __len__ method, which is required by Trainer for non-torch.utils.IterableDataset
+    # types. You should be able to fix this by calling dataset.with_format(type="torch") but that's not working for
+    # reason #1 (they use a local class which can't be pickled). You can work around this too (see ensure_format).
+    # 3. IterableDataset, when successfully pickled, tries to `open` the urls of the dataset as files (rather than
```
Comment (siddk) [inline src/corpora/auto.py:58]: This is HF Iterable Dataset? Another issue to upstream?

File: src/corpora/auto.py [hunk 5]
```diff
@@ -29,20 +34,41 @@ def get_auto_dataset(
...
+        return IterableDatasetDict({
+            k: StreamedDataset(dataset_id, dataset_name, k, str(paths["dataset"]), tokenizer, detokenizer, seq_len,
+                               stride) for k, _ in dataset.items()
+        })
+
     if "validation" not in dataset:
         assert "train" in dataset, "You must have train in dataset to make a validation dataset"
         # Create Dataset Split Cache Files
         train_fn, val_fn = [str(paths["dataset"] / dataset_id / f"{k}-split.hf") for k in ["train", "val"]]
+        # TODO: what to do for streaming?
```
Comment (siddk) [inline src/corpora/auto.py:71]: Probably raise an error here -- if streaming without a fixed validation set, either set a `no_validation` flag, or raise hell.

File: src/corpora/auto.py [hunk 6]
```diff
@@ -58,103 +84,98 @@ def get_auto_dataset(
...
     # Second, run straight-up tokenization
-    def tokenize(examples: Dict[str, List[str]]) -> BatchEncoding:
-        return tokenizer(examples["text"])
-
-    overwatch.info(f"Tokenizing Dataset via Multiprocessing with `{preprocessing_num_proc}` threads...")
+    if streaming:
+        tokenized_dataset = dataset.map(
+            _TokenizerFun(tokenizer),
+            batched=True,
+            remove_columns=list(next(iter(next(iter(dataset.values())))).keys()),
```
Comment (siddk) [inline src/corpora/auto.py:94]: This line is terrifying -- what's the double `next(iter(` doing?

**Review by siddk — COMMENTED** [review]
Yeah - this is scary, I'm reluctant to merge as is... wondering if we should bail on HF Datasets?

File: src/corpora/auto.py (hunk 2 above)
Comment (dlwh) [reply src/corpora/auto.py:125]: yeah i certainly think we should not try to rely on its streaming support, and possibly everything. What mosaic does is pretty gross but maybe this is what it needs to look like? https://github.com/stanford-crfm/composer/blob/pubmed/composer/datasets/c4.py#L127

File: src/corpora/auto.py (hunk 3 above)
Comment (dlwh) [reply src/corpora/auto.py:53]: yeah i will. I wanted to minimize to something that doesn't depend on mistral and haven't done it yet.

File: src/corpora/auto.py (hunk 4 above)
Comment (dlwh) [reply src/corpora/auto.py:58]: yes

File: src/corpora/auto.py (hunk 5 above)
Comment (dlwh) [reply src/corpora/auto.py:71]: yeah that's probably the right thing

File: src/corpora/auto.py (hunk 6 above)
Comment (dlwh) [reply src/corpora/auto.py:94]: yeah sorry, i need to refactor into a method, i just didn't b/c draft. it's just getting the keys of the the first example of the first split and saying that's the column names. It's identical in intention to the line in the other branch with column_names, but IterableDatasets don't expose column_names by default (which I understand, but it's still annoying)

Comment (siddk) [reply src/corpora/auto.py:94]: Got it - makes sense.

**Comment by dlwh** [comment]
I opened https://github.com/huggingface/datasets/issues/3950 and https://github.com/huggingface/datasets/issues/3951

**Comment by dlwh** [comment]
unnecessary with #138  and #139
